### ISSUE TITLE
CMD doesn't work when in array format for copying files

### ISSUE THREAD

**Issue by amir20** [docker/docker#30744]
**Description**

First a little backstory...
I have two containers. One container is a server using Caddy and another is my HTML files. Before, the HTML container was mounted using `volumes_from`. However, after a while, I realized that it is not updating when HTML files are updated. After a lot of research, and reading https://github.com/docker/compose/issues/3329, I realized that volumes doesn't do what I thought it does. It first copies from container to host, but second time it copies from host to container. (I am assuming this is expected to prevent loss of data?) 

Now -- the actual problem
It got me thinking, and I read a bunch of stackoverflow questions where instead of doing what I was doing, it would be better to use named volumes and copy data from a folder inside the container to the volume. When doing this, I think I have found a bug where `CMD` command doesn't work function properly. 

tldr:

- `CMD ["cp", "-rf", "/widget/*", "/var/www/html/"]`  does not work
- `CMD cp -rf /widget/* /var/www/html/` works

**Steps to reproduce the issue:**

My dockerfile looks like

```
FROM mhart/alpine-node:7

RUN npm install -g yarn

# most of this can be ignored 
RUN mkdir /tmp/app/

COPY package.json yarn.lock /tmp/app/
WORKDIR /tmp/app/
RUN yarn

COPY . /tmp/app/

RUN npm run build \
    && mv dist /widget \
    && cp test/index.html /widget/test.html \
    && rm -rf /tmp/app

# here is the command. /widget is created above and the volume is /var/www/html
CMD ["cp", "-rf", "/widget/*", "/var/www/html/"] 
```

My compose file is as follows: 

```
version: '2'
services:
  proxy:
    build: ./proxy
    restart: unless-stopped
    image: amir20/metum-proxy
    volumes:
        -   /root/.caddy
        - html:/var/www/html
  widget:
    build: ./widget
    image: amir20/metum-widget
    volumes:
      - html:/var/www/html/
volumes:
  html:
```

I run it by executing `docker-compose run --rm widget`

**Describe the results you received:**
When I execute this I receive `cp: can't stat '/widget/**': No such file or directory` error. Same thing if I use `command: ` in compose file. 

**Describe the results you expected:**
If I change the `CMD` to `CMD cp -rf /widget/* /var/www/html/` then it works as fine. So my expected result is for `CMD [...]` to work correctly with compose and/or with `[...]` format. 

**Additional information you deem important (e.g. issue happens only occasionally):**
I would actually prefer to use `ENTRYPOINT` and change the destination folder using `CMD`. But it doesn't work with docker-compose either. 

**Output of `docker version`:**

```
Client:
 Version:      1.13.0
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Wed Jan 18 16:20:26 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      1.13.0
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Wed Jan 18 16:20:26 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 9
 Running: 3
 Paused: 0
 Stopped: 6
Images: 163
Server Version: 1.13.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 201
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 2f7393a47307a16f8cee44a37b262e8b81021e3e
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.4-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952 GiB
Name: moby
ID: XB2I:YGQR:RLQV:ICKT:KK7M:23VI:HJTH:3WI3:6MAN:D5AP:XND7:QQWP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 42
 Goroutines: 47
 System Time: 2017-02-05T02:45:31.402797966Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Username: amir20
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

This is my first time reporting a bug on docker. I hope I followed the guideline.

**Comment by thaJeztah** [comment]
I think the issue here is that wild cards are expanded by the _shell_; when running

    CMD cp -rf /widget/* /var/www/html/

This command is run in a shell; and this is what's actually run;

    /bin/sh -c 'cp -rf /widget/* /var/www/html/'

When using the JSON notation, the shell is explicitly excluded, and `cp` is the command that's run. As a result, `cp` looks for a file literally called `*`.

I'm closing this issue, because this is not a bug, but feel free to continue the conversation

**Comment by amir20** [comment]
I was thinking that is the reason as well. So how do I run it correctly then? I need the json notation because I want the compose file to overwrite the volume destination. Is there a way to escape it?

**Comment by amir20** [comment]
Ps. Thanks for the quick response.

**Comment by willseward** [comment]
Can't do something like this?

`CMD ["/bin/bash", "-c", "cp -rf /widget/* /var/www/html/"]`

**Comment by amir20** [comment]
`CMD ["/bin/bash", "-c", "cp -rf /widget/* /var/www/html/"]` works but 

```

ENTRYPOINT ["/bin/sh", "-c", "cp -rf /widget/*"]
CMD ["/var/www/html/"]
```

Does not. Hmm. I think I'll have to be creative here. I might end up using `rsync` instead. 

Thanks, @thaJeztah, and @willseward. 

Can you (or anybody) please tell me if what I am doing is actually sound? As I understand correctly, volumes do not get updated when a container changes. Volumes are on way.  Is what I am doing the right way? Is there a better way to update data inside a volume from a container?

**Comment by willseward** [comment]
That's because `/var/www/html` is being passes as another word to `/bin/sh`. I believe the `cp` command needs to be passed to `bash` as a single parameter... I could be wrong, though...

To your question: I'm not acquainted well enough with your use-case or problem to offer any advice, but I would suggest looking into extracting your build step into another Dockerfile. Have a look at https://github.com/dnephin/dobi. It's basically Make for Docker.

**Comment by thaJeztah** [comment]
Using that as an entrypoint won't work, because the command will quit after it has completed, and the container will exit.

You'll need an entrypoint that performs the copy, and after it has completed, starts the containers main process. While not completely the same, you can have a look at the official WordPress image on Docker Hub, which first copies the files https://github.com/docker-library/wordpress/blob/d8e5577dcc7096d2b694dda433e2b5c7c4561a82/php5.6/apache/docker-entrypoint.sh#L45

and after the entrypoint script is completed,  starts the main process through `exec`; https://github.com/docker-library/wordpress/blob/d8e5577dcc7096d2b694dda433e2b5c7c4561a82/php5.6/apache/docker-entrypoint.sh#L193

**Comment by amir20** [comment]
> Using that as an entrypoint won't work, because the command will quit after it has completed, and the container will exit.

So I don't actually need to run anything in that container. It's just a container with javascript and static content. It would be ok if it quits. 

I ended up doing `ENTRYPOINT ["/usr/bin/rsync", "-rv", "/widget/"]` with docker-compose 

```
widget:
    build: ./widget
    image: amir20/metum-widget
    command: /var/www/html/
    volumes:
      - html:/var/www/html/
```

With output: 

```
$ docker-compose run --rm widget
sending incremental file list
b.js
b.js.gz
test.html

sent 32,189 bytes  received 73 bytes  64,524.00 bytes/sec
total size is 31,988  speedup is 0.99
```
 
This seems like the right thing to do and the best way to copy data from a container to a volume. I find it confusing that docker doesn't update the volume, but I guess it makes sense to defeat against data loss. 

Thanks @thaJeztah!

**Comment by thaJeztah** [comment]
> I  find it confusing that docker doesn't update the volume,

Volumes are designed to persist data created/generated/modified at runtime, so that you can update a container, without loosing those changes; updating the volume with the data from the image/container on start would result in overwriting that data, which is exactly what volumes try to prevent.

In your example; it looks like the data is static/read only. Is there a reason you don't package your HTML directory in the caddy image? In that case, you don't need a volume at all;

```Dockerfile
FROM your-caddy-proxy-image
COPY my-source-files /var/www/
```

Having said that; you may be interested in this proposal; https://github.com/docker/docker/issues/30449

**Comment by amir20** [comment]
> In your example; it looks like the data is static/read only. Is there a reason you don't package your HTML directory in the caddy image? In that case, you don't need a volume at all;

I have a caddy server where I proxy a bunch of subdomains to different docker containers. I had initially just a caddy container serving the files but then I added [other subdomains](https://github.com/amir20/metum.org/blob/master/proxy/Caddyfile), so I decided to just keep my proxy decoupled. I could have configured the proxy to another server but that just felt over complicated for my use case. 

I subscribed to #30449. Thanks!

**Comment by petegore** [comment]
Got the same behavior today : 
```
CMD ["cp", "-R", "/home/node/install/node_modules/*", "/home/node/app/node_modules"]
```
Returns a file not found error ; but 
```
CMD cp -R /home/node/install/node_modules/* /home/node/app/node_modules
```
Worked well.
